client/src/main/java/org/asynchttpclient/netty/NettyResponseFuture.java

```diff
@@ -454,6 +482,29 @@ public Object getPartitionKey() {
         return connectionPoolPartitioning.getPartitionKey(targetRequest.getUri(), targetRequest.getVirtualHost(), proxyServer);
     }
 
+    public void acquirePartitionLockLazily() throws IOException {
+        if (partitionKeyLock != null) {
+            return;
+        }
+
+        Object partitionKey = getPartitionKey();
+        channelManager.acquireChannelLock(partitionKey);
+        Object prevKey = partitionKeyLockField.getAndSet(this, partitionKey);
+        if (prevKey != null) {
+            // self-check
+
+            channelManager.releaseChannelLock(prevKey);
+            releasePartitionKeyLock();
+
+            throw new AssertionError("unreachable");
```

Comment:
```throw new IllegalStateException("Trying to acquire partition lock concurrently. Please report.");``` "Twice" is not proper word, because function is allowed to be called twice.